include/BidirectionalCoroutine.hpp

```diff
@@ -42,37 +45,111 @@ namespace com {
 						});
 					}
 				};
+
+				template<typename T>
+				using AlignedFor = typename std::aligned_storage<sizeof(T), alignof(T)>::type;
+
+				template<typename T>
+				class AlignedMaybeUninitializedDeleter {
+					std::unique_ptr<bool> initialized_;
+				public:
+					AlignedMaybeUninitializedDeleter() : initialized_(std::make_unique<bool>(false)) {}
+					AlignedMaybeUninitializedDeleter(const AlignedMaybeUninitializedDeleter&) = delete;
+					AlignedMaybeUninitializedDeleter& operator=(const AlignedMaybeUninitializedDeleter&) = delete;
+					AlignedMaybeUninitializedDeleter(AlignedMaybeUninitializedDeleter&& other) = default; 
+					AlignedMaybeUninitializedDeleter& operator=(AlignedMaybeUninitializedDeleter&&) = default;
+
+					void initialize() { *initialized_ = true; }
+					void reset() { initialized_ = std::make_unique<bool>(false); }
+					bool& initialized() { return *initialized_; }
+
+					void operator()(T *t) {
+						// I *think* this is correct, but could use a language lawyer
+						if(t) {
+							if(initialized_) {
+								if(*initialized_) {
+									t->~T();
+								}
+								auto storage = std::launder(reinterpret_cast<AlignedFor<T>*>(t));
+								delete storage;
+								initialized_ = nullptr;
+							} else {
+								throw std::runtime_error("An AlignedMaybeUninitializedDeleter may be used at most once");
+							}
+						}
+					}
+				};
+
+				template<typename T>
+				using UniqueMaybePtr = std::unique_ptr<T, AlignedMaybeUninitializedDeleter<T> >;
+
+				template<typename T>
+				UniqueMaybePtr<T> make_unique_uninitialized() {
+					return UniqueMaybePtr<T>(std::launder(reinterpret_cast<T*>(new AlignedFor<T>)));
+				}
+
+				template<typename T, typename ...Args>
+				T& emplaceMaybeUninitialized(T* t, bool& initialized, Args&& ...args) {
+					if(initialized) {
+						t->~T();
+					}
+					new (t) T(std::forward<Args>(args)...);
+					initialized = true;
+					return *t;
+				}
+
 			}
 			
 			template<class StackAlloc = boost::context::fixedsize_stack>
 			struct CoroutineContext {
 				using traits_type = typename StackAlloc::traits_type;
 				
-				template<class R, class ...Args> class BidirectionalCoroutine : protected BidirectionalCoroutine<void, Args...> {
-					R ret_;
+				template<class R, class ...Args>
+				class BidirectionalCoroutine : protected BidirectionalCoroutine<void, Args...> {
+					detail::UniqueMaybePtr<R> ret_;
 				protected:
 					using YieldVoid = typename BidirectionalCoroutine<void, Args...>::Yield;
+					using BidirectionalCoroutine<void, Args...>::next_;
 					
 				public:
 					class Yield : public YieldVoid {
-						using YieldVoid::handle_;
+						R* ret_; // Contractually, this must not be null
+						bool& rInit_; // The storage for this field is on the heap
```

Comment:
Please note that it is the Yield constructor that guarantees that this flag be on the heap.